src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1125,98 +1134,123 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-                        request.Content ??= new ByteArrayContent(Array.Empty<byte>());
-
-                        request.Content.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                    else
-                    {
-                        request.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                }
+                    // Create empty content so that we can send the entity-body header.
+                    _sendRequestMessage.Content ??= new ByteArrayContent(Array.Empty<byte>());
 
-                request.Headers.TransferEncodingChunked = SendChunked;
-
-                if (KeepAlive)
-                {
-                    request.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
+                    _sendRequestMessage.Content.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
                 }
                 else
                 {
-                    request.Headers.ConnectionClose = true;
+                    _sendRequestMessage.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
                 }
+            }
 
-                if (_servicePoint?.Expect100Continue == true)
-                {
-                    request.Headers.ExpectContinue = true;
-                }
+            if (_servicePoint?.Expect100Continue == true)
+            {
+                _sendRequestMessage.Headers.ExpectContinue = true;
+            }
+
+            _sendRequestMessage.Version = ProtocolVersion;
 
-                request.Version = ProtocolVersion;
+            _sendRequestMessage.Headers.TransferEncodingChunked = SendChunked;
+
+            if (KeepAlive)
+            {
+                _sendRequestMessage.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
+            }
+            else
+            {
+                _sendRequestMessage.Headers.ConnectionClose = true;
+            }
 
-                _sendRequestTask = async ?
-                    client.SendAsync(request, _allowReadStreamBuffering ? HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead, _sendRequestCts!.Token) :
-                    Task.FromResult(client.Send(request, _allowReadStreamBuffering ? HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead, _sendRequestCts!.Token));
+            _sendRequestMessage.Version = ProtocolVersion;
+            HttpCompletionOption completionOption = _allowReadStreamBuffering ?
+                            HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead;
+            _sendRequestTask = async ?
+                _httpClient.SendAsync(_sendRequestMessage, completionOption, _sendRequestCts!.Token) :
+                Task.Run(() => _httpClient.Send(_sendRequestMessage, completionOption, _sendRequestCts!.Token));
```

Comment:
Yes, this is basically removing the sync path of HttpWebRequest with this change, otherwise we will block the thread on `GetRequestStream` method call and request will never be completed.

Reply:
This sort of defeats the point of having a non-async path at all though.
We should try to do the work synchronously on the calling thread. If that's not possible and we have to block every time, then we might as well just use real async.